### PR TITLE
Batch metrics docs

### PR DESCRIPTION
**Pull request by knowlsie**
## Description of change
Docs and some helpers for batch metrics.

## Guide to reproduce test results.

`bazel test //resim/metrics:fetch_job_metrics_test`.

## Checklist:
- [x] I have self-reviewed this change.
- [x] I have tested this change.
- [x] This change is covered by tests that are already landed, or in this PR.

### CODE REVIEW

File: resim/metrics/fetch_job_metrics.py [hunk 1]
```diff
@@ -29,6 +34,42 @@ class JobInfo:
...
+    # Unpack the fetched metrics
+    unpacked_metrics_per_job = {}
+    for job in jobs:
+        metrics = metrics_protos[job.job_id]
+        metrics_data = metrics_data_protos[job.job_id]
+        unpacked_metrics_per_job[job.job_id] = unpack_metrics(
+            metrics=metrics,
+            metrics_data=metrics_data)
+
+    return unpacked_metrics_per_job
```
Comment (mikebauer) [inline resim/metrics/fetch_job_metrics.py:72]: Nice!

File: resim/metrics/fetch_job_metrics_test.py [hunk 2]
```diff
@@ -149,6 +200,58 @@ def test_fetch_job_metrics(self) -> None:
...
+@patch("resim.metrics.fetch_job_metrics.unpack_metrics",
+       new=_mock_unpack_metrics)
+@patch("resim_python_client.api.batches.list_jobs.sync",
+       new=_mock_list_job_ids_by_batch)
+@patch("resim_python_client.models.list_jobs_response_200.ListJobsResponse200",
+       MockListJobsResponse200)
+@patch("resim.metrics.fetch_job_metrics.UnpackedMetrics", MockUnpackedMetrics)
+@patch("resim_python_client.models.job.Job", MockJob)
+@patch("resim.metrics.fetch_job_metrics.mp.Metric", MockMetric)
+@patch("resim.metrics.fetch_job_metrics.mp.MetricsData", MockMetricsData)
```
Comment (mikebauer) [inline resim/metrics/fetch_job_metrics_test.py:218]: Nice mocking!

File: docs/resim/metrics/batch_metrics.md [hunk 3]
```diff
@@ -1,5 +1,57 @@
 # Batch metrics
 
-Our batch metrics system is currently functional but in an early phase where the interface is not fully finalized - as a result, the docs are a WIP. 
+Batch metrics are very similar to job metrics, but are computed after all the job metrics are computed, using the job-level metrics and metrics data as the "input."
 
-**If you are a customer and have any questions, feel free to reach out!**
+> Example Batch Metric 1: a weighted average of a job-level scalar metric, across all the jobs present.
+> Example Batch Metric 2: a histogram of a single series that is computed by all metrics jobs, after merging these series into one by appending.
```
Comment (mikebauer) [inline docs/resim/metrics/batch_metrics.md:6]: Formatting on this quote block seems a bit strange to me. Maybe add an extra line if you want each to show up on its own line? I've only looked at the GH markdown rendering though.

File: docs/resim/metrics/batch_metrics.md [hunk 4]
```diff
@@ -1,5 +1,57 @@
...
+
+Batch metrics can also be used to gate whether batches pass or fail, though this feature is currently still a WIP.
+
+## Batch metrics mode
+
+As mentioned in [the Metrics builds docs](./metrics_builds.md), batch metrics use the same single image as job metrics.
```
Comment (mikebauer) [inline docs/resim/metrics/batch_metrics.md:12]: [nit]
```suggestion
As mentioned in [the Metrics Builds docs](./metrics_builds.md), batch metrics use the same Docker image as job metrics. 
```

File: docs/resim/metrics/batch_metrics.md [hunk 5]
```diff
@@ -1,5 +1,57 @@
...
+> Example Batch Metric 1: a weighted average of a job-level scalar metric, across all the jobs present.
+> Example Batch Metric 2: a histogram of a single series that is computed by all metrics jobs, after merging these series into one by appending.
+
+Batch metrics can also be used to gate whether batches pass or fail, though this feature is currently still a WIP.
+
+## Batch metrics mode
+
+As mentioned in [the Metrics builds docs](./metrics_builds.md), batch metrics use the same single image as job metrics. 
+
+This image can differentiate whether it's computing in `job mode` or `batch mode`, based off whether the file `/tmp/resim/inputs/batch_metrics_config.json` is present. If it is present, it should be computing batch metrics.
```
Comment (mikebauer) [inline docs/resim/metrics/batch_metrics.md:14]:
```suggestion
This image can differentiate whether it's computing in "job mode" or "batch mode" based off whether the file `/tmp/resim/inputs/batch_metrics_config.json` is present. If it is present, it should be computing batch metrics.
```

File: docs/resim/metrics/batch_metrics.md [hunk 6]
```diff
@@ -1,5 +1,57 @@
...
+
+## Computing batch metrics
+
+The batch metrics config is a simple json with three fields: an auth token, an API URL, and a batch ID.
```
Comment (mikebauer) [inline docs/resim/metrics/batch_metrics.md:18]: [nit] Maybe make it more explicit that the worker provides this so customers don't think they need to do anything special to get the token?
```suggestion
The batch metrics config provided to the batch metrics runner on launch is a simple json with three fields: an auth token, an API URL, and a batch ID.
```

File: docs/resim/metrics/batch_metrics.md [hunk 7]
```diff
@@ -1,5 +1,57 @@
...
+  "batchID" : "7579affb-3e5b-4f02-871b-bf275aef67ee"
+}
+```
+
+These fields should be used to retrieve the job-level metrics and metrics data associated with a batch, and use these to compute batch-level metrics. We provide code to do this in [open-core](https://github.com/resim-ai/open-core), in combination with some code snippets below.
+
+First you can read the config in using the following snippet:
+
+```
+  with open(BATCH_METRICS_CONFIG_PATH, "r", encoding="utf-8") as metrics_config_file:
```
Comment (mikebauer) [inline docs/resim/metrics/batch_metrics.md:33]: [nit]
```suggestion
import json

  with open(BATCH_METRICS_CONFIG_PATH, "r", encoding="utf-8") as metrics_config_file:
```

File: docs/resim/metrics/batch_metrics.md [hunk 8]
```diff
@@ -1,5 +1,57 @@
...
+  api_url=metrics_config["apiURL"],
+  batch_id=metrics_config["batchID"],
+```
+
+Once these are loaded, you can download the metrics using our `fetch_job_metrics` Python package.
+
+```
+import resim.metrics.fetch_job_metrics as fjm
+
+job_to_metrics: Dict[uuid.UUID, UnpackedMetrics] = fjm.fetch_job_metrics_by_batch(token, api_url, uuid.UUID(batchID))
```
Comment (mikebauer) [inline docs/resim/metrics/batch_metrics.md:46]: Nice. The type hint here is informative.

File: docs/resim/metrics/batch_metrics.md [hunk 9]
```diff
@@ -1,5 +1,57 @@
...
+```
+
+The result maps job IDs to `UnpackedMetrics` - this is a simple `dataclass` with three fields:
+
+- metrics: List[Metric] - a list of all the metrics in that job
+- metrics_data: List[MetricsData] - a list of all the metrics data in that job
+- names: Set[str] - a set of all the names of Metrics and MetricsData present
+
+In other words, it very simply gives you all the metrics and metrics data associated with each job. You then use these metrics and data in order to compute and write your batch metrics - (just as you did for job metrics!) - by following the instructions in the [./metrics_writer.md](Metrics Writer docs). You write the output to the exact same place as before: `/tmp/resim/outputs/metrics.binproto`.
```
Comment (mikebauer) [inline docs/resim/metrics/batch_metrics.md:55]: The metrics_writer.md link appears to be the wrong way around.

**Review by mikebauer — CHANGES_REQUESTED** [review]
Looks great! Just a few small changes requested.

File: docs/resim/metrics/batch_metrics.md (hunk 5 above)
Comment (knowlsie) [reply docs/resim/metrics/batch_metrics.md:14]: Going to leave this one as is for consistency with previous docs.

File: docs/resim/metrics/batch_metrics.md [hunk 10]
```diff
@@ -1,5 +1,60 @@
...
-**If you are a customer and have any questions, feel free to reach out! If you aren't a customer yet and you're interesed in this feature and want to learn more, reach out to info@resim.ai and we can schedule some time together!**
+> Example Batch Metric 1: a weighted average of a job-level scalar metric, across all the jobs present.
+
+> Example Batch Metric 2: a histogram of a single series that is computed by all metrics jobs, after merging these series into one by appending.
+
+Batch metrics can also be used to gate whether batches pass or fail, though this feature is currently still a WIP.
+
+## Batch metrics mode
+
+As mentioned in [the Metrics builds docs](./metrics_builds.md), batch metrics use the same single Docker image as job metrics.
```
Comment (mikebauer) [inline docs/resim/metrics/batch_metrics.md:13]:
```suggestion
As mentioned in [the Metrics Builds docs](./metrics_builds.md), batch metrics use the same single Docker image as job metrics. 
```

File: docs/resim/metrics/batch_metrics.md (hunk 6 above)
Comment (mikebauer) [reply docs/resim/metrics/batch_metrics.md]: Looks good!

**Review by mikebauer — APPROVED** [review]

**Pull request merged**